legate/core/runtime.py

```diff
@@ -572,56 +585,58 @@ def prune_detachments(self) -> None:
-        self._num_pieces = runtime.core_context.get_tunable(
-            runtime.core_library.LEGATE_CORE_TUNABLE_NUM_PIECES,
-            ty.int32,
-        )
         self._min_shard_volume = runtime.core_context.get_tunable(
             runtime.core_library.LEGATE_CORE_TUNABLE_MIN_SHARD_VOLUME,
             ty.int64,
         )
 
-        if self._num_pieces == 0:
-            raise RuntimeError(
-                "No processors are available to run Legate tasks. Please "
-                "enable at least one processor of any kind. "
-            )
-
         self._launch_spaces: dict[
-            tuple[int, ...], Optional[tuple[int, ...]]
+            tuple[int, tuple[int, ...]], Optional[tuple[int, ...]]
         ] = {}
-        factors = list()
-        pieces = self._num_pieces
-        while pieces % 2 == 0:
-            factors.append(2)
-            pieces = pieces // 2
-        while pieces % 3 == 0:
-            factors.append(3)
-            pieces = pieces // 3
-        while pieces % 5 == 0:
-            factors.append(5)
-            pieces = pieces // 5
-        while pieces % 7 == 0:
-            factors.append(7)
-            pieces = pieces // 7
-        while pieces % 11 == 0:
-            factors.append(11)
-            pieces = pieces // 11
-        if pieces > 1:
-            raise ValueError(
-                "Legate currently doesn't support processor counts with prime "
-                "factors greater than 11"
-            )
-        self._piece_factors = list(reversed(factors))
+        self._piece_factors: dict[int, list[int]] = {}
+
         self._index_partitions: dict[
             tuple[IndexSpace, PartitionBase], IndexPartition
         ] = {}
         # Maps storage id-partition pairs to Legion partitions
         self._legion_partitions: dict[
             tuple[int, PartitionBase], Union[None, LegionPartition]
         ] = {}
-        self._storage_key_partitions: dict[int, PartitionBase] = {}
-        self._store_key_partitions: dict[int, PartitionBase] = {}
+        self._storage_key_partitions: dict[tuple[int, int], PartitionBase] = {}
+        self._store_key_partitions: dict[tuple[int, int], PartitionBase] = {}
+
+    def get_current_num_pieces(self) -> int:
+        return len(self._runtime.machine)
+
+    def get_piece_factors(self) -> list[int]:
+        num_pieces = self.get_current_num_pieces()
+        if num_pieces in self._piece_factors:
+            return self._piece_factors[num_pieces]
+
+        factors: list[int] = []
+
+        remaining = num_pieces
+
+        def factor_by(remaining: int, val: int) -> int:
+            while remaining % val == 0:
+                factors.append(val)
+                remaining = remaining // val
+            return remaining
+
+        remaining = factor_by(remaining, 2)
+        remaining = factor_by(remaining, 3)
+        remaining = factor_by(remaining, 5)
+        remaining = factor_by(remaining, 7)
+        remaining = factor_by(remaining, 11)
```

Comment:
If there were 3 or even 4 factors, maybe, but with this many (especially since the exc seems to indicate a possibility that the number may get bigger in the future):
```suggestion
        for factor in (2, 3, 5, 7, 11):
            remaining = factor_by(remaining, factor)
```

Reply:
https://github.com/nv-legate/legate.core/pull/457/commits/6d820fe7a60c0fac43bcc9413cd4daa26a4e14ba